source/use-cases/sensitive-data-encryption.txt

```diff
@@ -57,12 +57,222 @@ cubilia Curae; Nunc non interdum purus, ultricies laoreet tortor.
 B. Define a JSON Schema
 ~~~~~~~~~~~~~~~~~~~~~~~
 
-Proin non mi a felis luctus vulputate nec pharetra odio. Ut pretium
-scelerisque nulla, sed euismod nisi pellentesque sit amet. Proin mollis
-mauris eu libero ullamcorper rutrum. Pellentesque vitae enim sed magna
-accumsan maximus et quis mi. Ut vel laoreet turpis. Vivamus vitae odio
-sed arcu finibus pretium vel maximus nisi. Nullam gravida consequat
-porta.
+MongoDB uses JSON Schema to validate data that is written to a collection.
+Within the context of Field Level Encryption, JSONSchema is used to
+specify several configuration details.
+For instance, you can specify the form of encryption you want to use  -
+deterministic or random.
+You can also specify the BSON Type of the field you are encrypting. In
+addition, you can 
+specify the key you want to use. This is particularly important, because
+it means that you can use different keys for different fields.
+
+The following example goes over the creation of a schema for a collection
+that represents patient data
+in a medical office. Some fields use deterministic encryption, and others
+use random (probabilistic) encryption. 
+
+The structure for the schema will be as follows:
+
+.. code::
+
+    fullName: String
+    address: Encrypted String using random encryption
+    ssn: Encrypted Int using deterministic encryption
+    last4SSN: Int
+    patientInfo: Object
+    patientInfo.phone Encrypted Int using deterministic encryption
+    patientInfo.provider String
+    medicalRecords: Encrypted array of documents using random encryption
+
+Explanation of encryption choices
+``````````````````````````````````
+* ``address`` is randomly encrypted because it is more secure for strings to be randomly encrypted to prevent hackers from decrypting by building a cipher-dictionary
+* ``ssn`` is deterministically encrypted because deterministic encryption provides filtering ability 
+* ``patientInfo.phone`` is deterministically encrypted for the same reason as ssn, and also it provides an example of encryption with a nested document
+* ``medicalRecords`` are randomly encrypted because entire arrays are only encryptable with random encryption
+
+
+.. tabs::
+
+   tabs:
+
+     - id: java-jsonschema-generation
+       name: "Java"
+       content: |
+       
+        The following code creates a collection using the schema above and inserts
+        it into a database using MongoDB.
+        The createSecureClient is a method that calls getJSONSchema and passes it
+        a dataKey id in order to create a client that uses the JSONSchema returned
+        from getJSONSchema. However, the createSecureClient is abstracted from
+        this section for focus on the creation of schema. 
+
+        .. code-block:: java
```

Comment:
This code is nice as a template for implementing JSON Schema, but it's really long and leaves this section with little explanation of what's actually happening. We should explain what the `encrypt` field looks like generically and then demonstrate how we define it for each field to match our requirements.

Right now I have to parse through the code and the output to infer this stuff

Reply:
This might mean _not_ including the script as a single code block and instead splitting it up into individual pieces that we combine.

e.g.
- This is the generic schema for an encrypted field
  ```
  "<field name>": {
    "bsonType": "<BSON Type>",
    "encrypt": {
      "bsonType": "<BSON Type>",
      "algorithm": "<Encryption Algorithm>",
      "keyId": [
        {
          "$binary": {
            "base64": <keyId>,
            "subType": "04"
          }
        }
      ],
    }
  }
  ```
- You can generate the schema for an encrypted field using the following function...

Reply:
Also, can we use the Java BSON package to construct the Binary instead of manually doing it? (legit question idk if there's some limitation here)

Reply:
I agree that adding a simple example of what an encrypted field looks like would be helpful. I've added that to the most recent commit. 

As for your question: 
"Also, can we use the Java BSON package to construct the Binary instead of manually doing it?"
I don't know the answer to that, but it's unrelated to this exact section - which is specific to JSONSchema.

Reply:
If we build documents imperatively then using the official BSON package to construct BSON objects is very relevant and should be our preferred approach if feasible

Reply:
I'm wondering if the imperative, code-based approach we're taking here is actually the most effective approach. Right now we have two monolithic code blocks (one for the Java and one for the output). We also have content that focuses on the Java code but loses focus on actually explaining the "why" of each output schema.

This means I have to scroll back and forth a lot to map the context of each field to the schema. i.e. "which fields are encrypted?", "what encryption type does each field use?", and "what does the schema for a particular field/encryption type look like?" are all related but only the first two are grouped currently.

I think it would instead make sense to define the JSON Schema in an external `.json` file and then read & parse that file into our code. This has several advantages:

- It's way cleaner than imperative object creation
- It's consistent across all drivers (no need to write custom schema creation code for each driver)
- It makes it easier to copy-paste a completed schema that we provide into a file
- It allows us to put a small schema example in a code block right next to each field rather than separating the fields from the schema with a big code block. We can then give users a link to see/download the full JSON Schema without taking up a ton of screen real estate.

Let's sync on this in person / over slack. cc: @ccho-mongodb

Reply:
^ Essentially, the size of the "address" schema example in the most recent version is ideal - we should aim for all of the fields to use bite-size code blocks like that. We can also add line highlights to point out the key things that we mention in accompanying text.

Reply:
I agree that having large monolithic code blocks causes a lot of annoying scrolling. This guide (meaning the entire page, not just JSON schema section) is meant to be a complete example of an implementation of field-level encryption that could happen in real life. The one thing that I think can help in terms of breaking up the monolithic nature of the code blocks is separating the java code. So for instance, creating one code block with createEncryptedFieldType and then an explanation of what that results in, and then another code block of the method createStandardFieldType and then an explanation of that and then the final code block of getJSONSchema which uses the other functions to create the entire schema and the results of the schema in JSON form (so the second monolithic code-block, the output, would remain)


As for making the JSON schema an external file, I’ve considered this. Java generally doesn’t have any “great” ways of creating JSON like javascript does, so even if we pivot to another option, I am certain it will be both time-consuming and non-fruitful in terms of adding clarity or decreasing bulk to the page. 

So the three original options were:
External file 
As a string
Using a JSONObject creation import (our current use-case)

Number two would be the most absurd option unless the schema we want was really limited and small (this is the route the java drivers team took in their doc, which made sense for them since their example was just one field). Number one is usable but the issue here is that a JSON file can’t be dynamic. I think most readers will probably use a JSONObject creation library of some sort (there’s at least a few out there). In my opinion, we should keep using number 3 with the caveat that we split the huge code chunk into three code chunks per method that each has an explanation. 

Also note: I have a bias to number 3 because I’ve just spent the past week writing and rewriting the doc using the JSONObject method.